Genesis DAO 2nd Milestone

deep-ink-ventures opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x In case of acceptance, the payment will be transferred to the BTC/ETH/fiat account provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/1383

dsm-w3f commented:
@deep-ink-ventures thank you for the milestone submission. Please see the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/genesisdao_2_dsm-w3f.md) and provide proper answers and fixes. After that, let me know to continue with the evaluation.

deep-ink-ventures commented:
@dsm-w3f thank you!!

## Licence
We added the [licence](https://github.com/deep-ink-ventures/genesis-dao-service/blob/main/LICENCE).

## Documentation

Added for the [node](https://github.com/deep-ink-ventures/genesis-dao-node/blob/main/README.md#docker-setup), [frontend](https://github.com/deep-ink-ventures/genesis-dao-frontend#docker-setup) and [service](https://github.com/deep-ink-ventures/genesis-dao-service#quickstart-using-docker).


## Docker

> Node Logs:

We're not able to reproduce this error, the docker container runs fine for us (in fact [this](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fnode.genesis-dao.org#/explorer) is using that docker file).

However, this is a common error if one has an outdated base image. Since the last milestone we've upgraded to [0.9.40](https://github.com/paritytech/polkadot/issues/6888) and you might to explicitly run
`docker pull paritytech/ci-linux:production` to prune the docker cache and enforce the latest rust version.

> Service Logs:

We've simplified the setup, you should now be able to copy & paste through the [Installation / Quickstart using Docker](https://github.com/deep-ink-ventures/genesis-dao-service#installation). Please be aware that this is by default syncing to the testnet, so depending on the current block number the listener may take A WHILE to catch up (you can see it in the logs). You can either just live with the API not being up to date or point it to a local node (by adjusting the `.env` file) or ping me and I'll reset the testnet for you so it's easier to catch up.

## Manual Testing


> I ran the frontend without problems, but when I tried to create a dao got this message "Sorry you need at least 10 DOT tokens to create a DAO. You will get them back if you destroy the DAO.". Even that I get more balance for the connected account, the error msg remains.

Yes, that's our bad. We've changed the [chainspec](https://github.com/deep-ink-ventures/genesis-dao-node/commit/cd5a4f2a544a570e5469d7f034c97f3b2eaa2ed2) and had subsequent deployment and you tested in the middle. We (again :-)) promise to not touch main anymore, but wanted to have the same decimals as dot for the intermediate native currency. This should work now.

> I tried to run the services using docker and got the error mentioned above. So I tried to run without docker and got this error:

That's likely a wrong configuration. We clarified the documentation on how to run without docker.

dsm-w3f commented:
@deep-ink-ventures thank you for the improvements and fixes. I checked again and some errors remain. I was able to test the application running the frontend locally and pointing it to the hosted service but I couldn't start it with all parts locally. See details in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/genesisdao_2_dsm-w3f.md). Let me know when I can continue this evaluation.

deep-ink-ventures commented:
Thanks again @dsm-w3f .

> This time, the listener service started but still have errors with worker one. I tried to visit http://127.0.0.1:8000/redoc/ and got a blank page.

We where able to reproduce this and made the [static file discovery](https://github.com/deep-ink-ventures/genesis-dao-service/commit/9143f7c8fd0e8bd979da7718826f80fcecc600a8#diff-27aecacbb2748a2bfdde9016eb13db9608d7fa8808cadbc4a3be7398fcf8d990R43) explicit. 

This solves the "blank page", but we where not able to reproduce the error the worker gives you in the logs. If the error persists, can you please try to `chmod` your directory recursively (especically the static folder) to `777` (just for debugging purposes)?

I'm as well happy to jump on a call with you to fix this.

> I tried without Docker, and this time I received this error when I ran make start-dev:

The `match` syntax was introduced in python 3.10, we require 3.11+; this error is likely due to an outdated python version.

> But I noticed the page with Review Details is missing.

Yes, we removed the review page because of the feedback we got from testers: We are incrementally commit the changes to the chain during setup and users thought that the "review" means nothing has been done so far. We replaced it with a "return" functionality, if you interrupt DAO creation the onboarding flow takes you back to the right screen.

> And another message as well "Sorry, you are not the admin of [Dao Name]". This this last one I think should't be presented.

We are still finalising the dashboard page for the next milestone, but this is irritating I agree (technically you are no longer admin, but signer on the multisig of the admin) and we applied a fix.

dsm-w3f commented:
@deep-ink-ventures thank you for the answer and fixes. I tried again to run the system using a local environment but without success. Please see details in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/genesisdao_2_dsm-w3f.md). The ideal scenario for a system with multiple parts is to have a docker compose that start all parts of it and connect them.

deep-ink-ventures commented:
Thanks, @dsm-w3f.

When you run `docker compose` from different repos they create different networks that need to be wired. I agree, though, that this is not cool when you want to test things out, hence we unified everything in the [node repo](https://github.com/deep-ink-ventures/genesis-dao-node/commit/de45d6352f8a873a664a062c782e7a0540b7c40a).

When you run 

```
docker compose up
docker compose build
```

within the node repo, the entire system spins up.

As for the other error, I am still not able to reproduce it, but the Docker spin up for the node is now cleaner as well, so ideally it just vanishes.

dsm-w3f commented:
@deep-ink-ventures thank you for the improvements and fixes. I still having problems to run the system locally. Please see details in the [evaluation document](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/genesisdao_2_dsm-w3f.md) for details. Let me know when I can continue this evaluation.

deep-ink-ventures commented:
Thanks @dsm-w3f!

> I tested the docker and got again the same error, but I was able to solve this by modifying the files

This must be a caching error on your system. [Our Dockerfile](https://github.com/deep-ink-ventures/genesis-dao-node/blob/main/Dockerfile#L1) inherits from parity's original image that in turn already [installs the nightly](https://github.com/paritytech/scripts/blob/master/dockerfiles/ci-linux/Dockerfile#L26) and is already instaling the [wasm toolchain](https://github.com/paritytech/scripts/blob/master/dockerfiles/ci-linux/Dockerfile#L30). Happy to jump on a call and find out the root of that, but I'm still not able to reproduce this error.

> And with this, Docker started the node, service, and front end. But it doesn't seem to be connected. I tried to connect my wallet and received the error "n.balance is undefined", seems to be the same error I received trying without docker "account.balance is undefined". I checked the containers and genesis-dao-node-listener-1 has some problems.

This is a race condition. The migrations haven't run yet in the listener. I've added [the migrations](https://github.com/deep-ink-ventures/genesis-dao-service/commit/060da67142652c78d6a8d635eb8d79faa92c3a42) to all systems and [we made them concurrent](https://github.com/deep-ink-ventures/genesis-dao-service/blob/main/core/management/commands/save_migrate.py#L8-L9).

>  I got this error using the command pre-commit install from make build.

Did you follow the [prerequisites](https://github.com/deep-ink-ventures/genesis-dao-service#prerequisites) that are required for running w/o docker, specifically the [sqlite](https://www.w3resource.com/sqlite/sqlite-download-installation-getting-started.php) guide?

[This error](https://stackoverflow.com/questions/1210664/no-module-named-sqlite3) is a result of this.

Generally speaking: You seem to run into some race conditions and caching problems on a rather complicated setup of different services. Wdyt about jumping on a quick call to iron that out instead of going back and forth here, as we are not able to reproduce those problems locally and are more or less guessing.

dsm-w3f commented:
@deep-ink-ventures thank you for your answer. I'll double-check taking your considerations next week. Please also check if there is no cache problem on your side when downloading and building the images. I saw other grants having this exact same build problem and they solved adding some instructions to the dockerfile as I did. Regarding the possible race condition, there are ways to control that with docker compose if I remember well, there are dependencies between containers and health checks that could be implemented to control that. Race conditions should be prevented by design in the application. I'll try again to solve the dependency with Sqlite although I think this should be solved by the software installation instructions or docker containers. Expect to hear news from me by Tuesday of the next week.

deep-ink-ventures commented:
Enjoy your weekend! 

> Please also check if there is no cache problem on your side when downloading and building the images. I saw other grants having this exact same build problem and they solved adding some instructions to the dockerfile as I did.

I will try on a fresh computer as well. Happy to add your lines if it fixes things, from my understanding those are already run in the image it inherits, but I will have another look as well.

> Race conditions should be prevented by design in the application.

Yes! We added the fixed after your last evaluation. We didn't have that before, but it should now be fixed.

>  I'll try again to solve the dependency with Sqlite although I think this should be solved by the software installation instructions or docker containers

SQLLite is just a fallback if you don't use the postgres container in docker. This was a bit misleading in the README, the prerequisites including SQLLite are not required when using docker. I moved them accordingly in the docs:


https://github.com/deep-ink-ventures/genesis-dao-service/commit/b276c1fa62991d5425abb3f1c56e6b62a13201da

deep-ink-ventures commented:
Update: I was able to reproduce your error on a fresh ec2 instance and indeed your three lines are fixing it - i added them [here](https://github.com/deep-ink-ventures/genesis-dao-node/commit/19e47ade0ee53db3b75c32f3d0aa519e792a6d3d). Together with the aforementioned race condition fix it now runs fine for me on a virgin ec2 instance.

dsm-w3f commented:
@deep-ink-ventures thank you for the improvements and fixes. I tried again using docker compose and all containers spin up properly. However, when I tried to access the frontend application the frontend container gives this error and stop running. Any hint on that? Seems to be a connection problem for me.

```
genesis-dao-node-web-1       | node:internal/event_target:1012
genesis-dao-node-web-1       |   process.nextTick(() => { throw err; });
genesis-dao-node-web-1       |                            ^
genesis-dao-node-web-1       | 
genesis-dao-node-web-1       | TypeError: Cannot read properties of null (reading 'length')
genesis-dao-node-web-1       |     at WebSocket.value (file:///usr/src/app/node_modules/@polkadot/rpc-provider/ws/index.js:213:91)
genesis-dao-node-web-1       |     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:737:20)
genesis-dao-node-web-1       |     at WebSocket.dispatchEvent (node:internal/event_target:679:26)
genesis-dao-node-web-1       |     at fireEvent (/usr/src/app/node_modules/next/dist/compiled/undici/index.js:3:4860)
genesis-dao-node-web-1       |     at websocketMessageReceived (/usr/src/app/node_modules/next/dist/compiled/undici/index.js:3:5188)
genesis-dao-node-web-1       |     at ByteParser.run (/usr/src/app/node_modules/next/dist/compiled/undici/index.js:3:3261)
genesis-dao-node-web-1       |     at ByteParser._write (/usr/src/app/node_modules/next/dist/compiled/undici/index.js:3:967)
genesis-dao-node-web-1       |     at writeOrBuffer (node:internal/streams/writable:392:12)
genesis-dao-node-web-1       |     at _write (node:internal/streams/writable:333:10)
genesis-dao-node-web-1       |     at Writable.write (node:internal/streams/writable:337:10)
genesis-dao-node-web-1       | 
genesis-dao-node-web-1       | Node.js v18.16.0

```

deep-ink-ventures commented:
We're getting there!

Yes, this error is familiar. We've added a few changes to prevent this. Most prominently, we didn't copy the `yarn.lock` into the docker container which creates random behaviour like this.

It should be fixed now. You should be able to just `docker compose build web` from the node and then `docker compose up` again.

dsm-w3f commented:
@deep-ink-ventures thank you for the improvements and fixes. Now I was able to run the system properly using docker. The milestone is accepted. I'll forward your invoice internally and the payment should take place within two weeks. Great job!
